design/custom-request-timeout.md

```diff
@@ -0,0 +1,256 @@
+# Executive Summary
+
+**Status**: _Draft_
+
+This document describes the design of a new resource in IngressRoute for custom request timeout and retries. 
+
+This new resource will be integrated into Contour 0.11.
+
+# Goals
+
+- Generic resource name for configuring custom request timeouts per IngressRoute
+- Generic resource name for configuring custom retry attempts per IngressRoute
+
+# Non-goals
+
+- None
+
+# Background
+
+Contour supports custom request timeout and custom retry attempts via Ingress. The same functionality has been
+requested via IngressRoute as well.
+
+# High-level design
+
+At a high level, this document proposes the changes in the IngressRoute's CRD to also support custom request
+timeouts and retries. It covers:
+
+- Fields added for timeout and retries in the YAML file.
+- A high-level coverage on the struct changes required in the code.
+- Arguments passed for Envoy API calls.
+
+## Proposed YAML fields
+
+```yaml
+apiVersion: contour.heptio.com/v1beta1
+kind: IngressRoute
+metadata:
+    name: github.com
+    namespace: prod
+spec:
+    virtualhost:
+        fqdn: www.github.com
+        routes:
+        - match: /contour
+          # the timeout applied to the requests in this route
+          timeoutPolicy:
+              # the request timeout applied on this route, it is the maximum time that a client will await
+              # a response to its request.
+              request: 500ms
+              # the maximum time that an idle connection is kept open when there are no active requests
+              idle: 2s
+          retryPolicy:
+              # the number of retries it should make on failure
+              maxRetries: "7"
+              # the conditions for which it should reattempt
+              onStatusCodes:
```

Comment:
Your assumptions are correct.

I believe the user should be able to specify `5xx` and `4xx` in the YAML file. When some of the "xx" is included, the contour should map it to `"5xx"` or `"retriable-4xx"` respectively against `retry_on` for Envoy.  If `50x` or `40x` is specified, contour should expand that to `500,501,502..,509` and `400,401,402...,409` just like the specific status codes, which will be mapped to `retriable_status_codes`.

Reply:
based on the recent feedback for keeping ingressroute generic for all controllers, ingressroute for now will only have `50x` or {500->509} based status codes only.

Reply:
`retry_on` provides difference schematics than `retriable_status_codes`. For example:

> 5xx
> Envoy will attempt a retry if the upstream server responds with any 5xx response code, or does not respond at all (disconnect/reset/read timeout). (Includes connect-failure and refused-stream) 

5xx retry policy consists of cases where upstream returns 5xx in response as well as those with no response code at all.

But I understand that probably we don't want to tie Contour with Envoy too closely per issues/587.

Reply:
Sorry for not making it clear in the design doc. A user can specify

```
  retryPolicy:
    onStatusCodes:
      - "505"
      - "509"
      - "501"
```

Since contour sees them all as numbers, it will set "retriable-status-codes" for `retry_on`, and actually fill in uint32 type status codes for `retriable_status_codes`. For anyother ingress controller using ingressroute, they should treat them as status codes to perform retry.

However, if contour sees anything like the following:

```
retryPolicy:
  onStatusCodes:
    - "50x"
    - "503"
```

Contour (or any ingress controller) sees that "50x" has been specified and it should expand to [500->509]. To configure for envoy, since we don't want to retry`5xx` (ie status codes >= 510 which is outside user's configuration), we will repeat the step as in previous example of setting "retriable-status-codes" for `retry_on` and then setting uint32 type status codes to `retriable_status_codes`.